src/term/parser/parser.rs

```diff
@@ -205,47 +205,101 @@ where
   choice((pat_nam, pat_tup))
 }
 
-fn definition<'a, I>() -> impl Parser<'a, I, (Name, Definition), extra::Err<Rich<'a, Token>>>
+fn rule_pat<'a, I>() -> impl Parser<'a, I, RulePat, extra::Err<Rich<'a, Token>>>
 where
   I: ValueInput<'a, Token = Token, Span = SimpleSpan>,
 {
-  let lhs = choice((name(), name().delimited_by(just(Token::LParen), just(Token::RParen))));
+  recursive(|rule_pat| {
+    let var = name().map(RulePat::Var).boxed();
+
+    let ctr = name()
+      .then(rule_pat.clone().repeated().collect())
+      .map(|(nam, xs)| RulePat::Ctr(nam, xs))
+      .delimited_by(just(Token::LParen), just(Token::RParen))
+      .boxed();
+
+    choice((var, ctr))
+  })
+}
+
+fn rule<'a, I>() -> impl Parser<'a, I, (Name, Rule), extra::Err<Rich<'a, Token>>>
+where
+  I: ValueInput<'a, Token = Token, Span = SimpleSpan>,
+{
+  let def_name =
+    name().try_map(|Name(nam), span| if nam != "data" { Ok(Name(nam)) } else { Err(Rich::custom(span, "")) });
+  let lhs = def_name.then(rule_pat().repeated().collect()).boxed();
+  let lhs = choice((lhs.clone(), lhs.clone().delimited_by(just(Token::LParen), just(Token::RParen))));
 
   lhs
     .then_ignore(just(Token::Equals))
     .then(term())
-    .map(|(name, body)| (name, Definition { def_id: DefId(0), body }))
+    .map(|((name, pats), body)| (name, Rule { def_id: DefId(0), pats, body }))
 }
 
-fn book<'a, I>() -> impl Parser<'a, I, DefinitionBook, extra::Err<Rich<'a, Token>>>
+fn datatype<'a, I>() -> impl Parser<'a, I, (Name, Adt), extra::Err<Rich<'a, Token>>>
 where
   I: ValueInput<'a, Token = Token, Span = SimpleSpan>,
 {
-  fn definitions_to_book(
-    defs: Vec<((Name, Definition), SimpleSpan)>,
-    _span: SimpleSpan,
-    emitter: &mut Emitter<Rich<Token>>,
-  ) -> DefinitionBook {
-    let mut book = DefinitionBook::new();
-
-    // Check for repeated defs (could be rules out of order or actually repeated names)
-    for ((name, Definition { body, .. }), def_span) in defs {
-      if !book.def_names.contains_name(&name) {
-        let def_id = book.def_names.insert(name);
-        book.defs.insert(def_id, Definition { def_id, body });
-      } else {
-        let (start, end) = (def_span.start, def_span.end);
-        let span = SimpleSpan::new(start, end);
-        emitter.emit(Rich::custom(span, format!("Repeated definition '{}'", name)));
+  let arity_0 = name().map(|nam| (nam, 0));
+  let arity_n = name()
+    .then(name().repeated().collect::<Vec<_>>())
+    .delimited_by(just(Token::LParen), just(Token::RParen))
+    .map(|(nam, args)| (nam, args.len()));
+  let ctr = arity_0.or(arity_n);
+
+  let data =
+    name().try_map(|Name(nam), span| if nam == "data" { Ok(()) } else { Err(Rich::custom(span, "")) });
```

Comment:
Can you make a function for parsing a soft keyword?  So this whole line, but you receive `"data"` as an argument to the function. So here we just call `let data = soft_keyword("data");`. This could then just be inlined in the line below, like

```
  soft_keyword("data")
    .ignore_then(name())
    .then_ignore(just(Token::Equals))
    .then(ctr.separated_by(just(Token::Or)).collect::<Vec<(Name, usize)>>())
    .map(|(name, ctrs)| (name, Adt { ctrs: ctrs.into_iter().collect() }))
```